File: go/list-ops/list_ops.go
package listops

type IntList []int
type binFunc func(int, int) int
type unaryFunc func(x int) int
type predFunc func(n int) bool

// Append will append the list appendList onto list l and returns a new list
func (l IntList) Append(appendList IntList) IntList {
	resultList := make(IntList, len(l)+len(appendList))
	for i := 0; i < len(l); i++ {
		resultList[i] = l[i]
	}
	for j := 0; j < len(appendList); j++ {
		resultList[j+len(l)] = appendList[j]
	}
	return resultList
}

func (l IntList) Concat(lists []IntList) IntList {
	var resultList = l
	for _, list := range lists {
		resultList = append(resultList, list...)
	}
	return resultList
}

// Filter removes items in list that do not satisfy the filterfunc
func (l IntList) Filter(filterFunc predFunc) IntList {
	resultList := make(IntList, 0)
	for i := 0; i < len(l); i++ {
		if filterFunc(l[i]) {
			resultList = append(resultList, l[i])
		}
	}
	return resultList
}

func (l IntList) Length() int {
	var length int
	for i, _ := range l {
		length = i + 1
	}
	return length
}

func (l IntList) Map(mapFunc unaryFunc) IntList {
	resultList := make(IntList, len(l))
	for i := 0; i < len(l); i++ {
		resultList[i] = mapFunc(l[i])
	}
	return resultList
}

func (l IntList) Foldl(foldFunc binFunc, initial int) int {
	var accumulator = initial
	for i := 0; i < len(l); i++ {
		accumulator = foldFunc(accumulator, l[i])
	}
	return accumulator
}

func (l IntList) Foldr(foldFunc binFunc, initial int) int {
	var accumulator = initial
	for i := len(l) - 1; i > -1; i-- {
		accumulator = foldFunc(l[i], accumulator)
	}
	return accumulator
}

func (l IntList) Reverse() IntList {
	resultList := make(IntList, len(l))
	var resultIndex int
	for i := len(l) - 1; i > -1; i-- {
		resultList[resultIndex] = l[i]
		resultIndex++
	}
	return resultList
}

File: javascript/list-ops/list-ops.js
export class List {
    constructor(values = []) {
        this.values = values
    }

    // append adds a new list to existing list and return a new list
    append(list) {
        return new List([...this.values, ...list.values])
    }

    // concatenate (given a series of lists, combine all items in all lists into one flattened list)
    concat(listArray) {
        return this.append(new List(listArray.foldl((acc, item) => [...acc, ...item.values], [])))
    }

    // given a predicate and a list, return the list of all items for which predicate(item) is True)
    filter(predicate) {
        return new List(this.foldl((acc, element) => predicate(element) ? [...acc, element] : [...acc], []))
    }

    // given a function and a list, return the list of the results of applying function(item) on all items)
    map(func) {
        return new List(this.foldl((acc, element) => [...acc, func(element)], []))
    }

    // length returns the size of the list
    length() {
        return this.foldl(length => ++length, 0)
    }

    // fold is a generic helper function that works the same as foldl that accepts array as input
    fold(func, acc, arr) {
        const [first, ...rest] = arr
        if (first === undefined)
            return acc
        acc = func(acc, first)
        return this.fold(func, acc, rest)
    }

    // foldl (reduce) each item into the accumulator from the left using function(accumulator, item))
    foldl(func, accumulator) {
        return this.fold(func, accumulator, this.values)
    }

    // foldr (given a function, a list, and an initial accumulator, 
    // fold (reduce) each item into the accumulator from the right using function(item, accumulator))
    foldr(func, accumulator) {
        return this.reverse().foldl(func, accumulator)
    }

    // reverse (given a list, return a list with all the original items, but in reversed order)
    reverse() {
        return new List(this.foldl((acc, element) => [element, ...acc], []))
    }
}